docs/structured-streaming-programming-guide.md

```diff
@@ -1848,12 +1848,137 @@ Additional details on supported joins:
 
 - As of Spark 2.4, you can use joins only when the query is in Append output mode. Other output modes are not yet supported.
 
-- As of Spark 2.4, you cannot use other non-map-like operations before joins. Here are a few examples of
-  what cannot be used.
+- You cannot use mapGroupsWithState and flatMapGroupsWithState before and after joins.
 
-  - Cannot use streaming aggregations before joins.
+In append output mode, you can construct a query having non-map-like operations before/after join.
```

Comment:
`non-map-like operations` sounds not clear to me. Can you add some examples like:

```
In append output mode, you can construct a query having non-map-like operations, e.g. ..., ..., ..., before/after join.
```

Reply:
Sounds better. I'll list some operations here.